In [87]:
import numpy as np
import pandas as pd

In [122]:
df = pd.read_csv("C:/Users/Hp/Desktop/sarathi/Day11/dataset.csv")
df

,Series_reference,Period,Data_value,Suppressed,STATUS,UNITS,Magnitude,Subject,Group,Series_title_1,Series_title_2,Series_title_3,Series_title_4,Series_title_5
0,BDCQ.SEA1AA,2011.06,80078.0,NaN,F,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Actual,NaN,NaN
1,BDCQ.SEA1AA,2011.09,78324.0,NaN,F,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Actual,NaN,NaN
2,BDCQ.SEA1AA,2011.12,85850.0,NaN,F,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Actual,NaN,NaN
3,BDCQ.SEA1AA,2012.03,90743.0,NaN,F,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Actual,NaN,NaN
4,BDCQ.SEA1AA,2012.06,81780.0,NaN,F,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Actual,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
494,BDCQ.SEA1DS,2019.12,19634.0,NaN,R,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Electricity, Gas, Water and Waste Services",Seasonally adjusted,NaN,NaN
495,BDCQ.SEA1DS,2020.03,19770.0,NaN,R,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Electricity, Gas, Water and Waste Services",Seasonally adjusted,NaN,NaN
496,BDCQ.SEA1DS,2020.06,19791.0,NaN,R,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Electricity, Gas, Water and Waste Services",Seasonally adjusted,NaN,NaN
497,BDCQ.SEA1DS,2020.09,19886.0,NaN,R,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Electricity, Gas, Water and Waste Services",Seasonally adjusted,NaN,NaN


In [89]:
df.describe()

,Period,Data_value,Suppressed,Magnitude,Series_title_4,Series_title_5
count,499.000000,361.000000,0.0,499.0,0.0,0.0
mean,2016.505952,83802.180055,NaN,0.0,NaN,NaN
std,3.305866,84137.334139,NaN,0.0,NaN,NaN
min,2011.060000,5060.000000,NaN,0.0,NaN,NaN
25%,2014.030000,6004.000000,NaN,0.0,NaN,NaN
50%,2016.120000,81471.000000,NaN,0.0,NaN,NaN
75%,2019.105000,200677.000000,NaN,0.0,NaN,NaN
max,2022.090000,228672.000000,NaN,0.0,NaN,NaN


In [90]:
df.head()

,Series_reference,Period,Data_value,Suppressed,STATUS,UNITS,Magnitude,Subject,Group,Series_title_1,Series_title_2,Series_title_3,Series_title_4,Series_title_5
0,BDCQ.SEA1AA,2011.06,80078.0,NaN,F,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Actual,NaN,NaN
1,BDCQ.SEA1AA,2011.09,78324.0,NaN,F,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Actual,NaN,NaN
2,BDCQ.SEA1AA,2011.12,85850.0,NaN,F,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Actual,NaN,NaN
3,BDCQ.SEA1AA,2012.03,90743.0,NaN,F,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Actual,NaN,NaN
4,BDCQ.SEA1AA,2012.06,81780.0,NaN,F,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Actual,NaN,NaN


In [26]:
df.tail()

,Series_reference,Period,Data_value,Suppressed,STATUS,UNITS,Magnitude,Subject,Group,Series_title_1,Series_title_2,Series_title_3,Series_title_4,Series_title_5
494,BDCQ.SEA1DS,2019.12,19634.0,NaN,R,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Electricity, Gas, Water and Waste Services",Seasonally adjusted,NaN,NaN
495,BDCQ.SEA1DS,2020.03,19770.0,NaN,R,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Electricity, Gas, Water and Waste Services",Seasonally adjusted,NaN,NaN
496,BDCQ.SEA1DS,2020.06,19791.0,NaN,R,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Electricity, Gas, Water and Waste Services",Seasonally adjusted,NaN,NaN
497,BDCQ.SEA1DS,2020.09,19886.0,NaN,R,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Electricity, Gas, Water and Waste Services",Seasonally adjusted,NaN,NaN
498,BDCQ.SEA1DS,2020.12,20213.0,NaN,R,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Electricity, Gas, Water and Waste Services",Seasonally adjusted,NaN,NaN


In [33]:
my_seperate_df=df.groupby("Series_reference").get_group("BDCQ.SEA1AA")

In [13]:

#df.groupby("subject")["score"].mean()


In [36]:
my_seperate_df['Data_value'].mean()
my_seperate_df['Data_value'].describe()

count        46.000000
mean      92019.478261
std        7028.150017
min       78324.000000
25%       86969.000000
50%       91603.000000
75%       98750.250000
max      102683.000000
Name: Data_value, dtype: float64

In [147]:
students = pd.DataFrame({
    "id": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    "name": [
        "Alice",
        "Bob",
        "Charlie",
        "David",
        "Emma",
        "Frank",
        "Grace",
        "Henry",
        "Isabella",
        "Jack"
    ],
    "subjects": [
        "Math",
        "Science",
        "English",
        "Computer",
        "Physics",
        "Chemistry",
        "Biology",
        "History",
        "Economics",
        "Geography"
    ]
}
    )

scores = pd.DataFrame({
    "id": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    "score": [85, 92, 78, 88, 95, 81, 76, 89, 91, 84],
    "subject": [
        "Math",
        "Science",
        "English",
        "Computer",
        "Physics",
        "Chemistry",
        "Biology",
        "History",
        "Economics",
        "Geography"
    ]
}
)

pip freeze >requirements.txt

In [148]:
#Exactly like sql join
merged = pd.merge(students,scores,on = "id")
#Only id gets combined and all others are joined 
print(merged.head())
merged["score"].unique()




   id     name  subjects  score   subject
0   1    Alice      Math     85      Math
1   2      Bob   Science     92   Science
2   3  Charlie   English     78   English
3   4    David  Computer     88  Computer
4   5     Emma   Physics     95   Physics


array([85, 92, 78, 88, 95, 81, 76, 89, 91, 84])

In [57]:
grouped = merged.groupby("score").get_group(84)
grouped
print(merged["score"]>40)
greater_thanforty = merged[merged["score"]>40]
greater_thanforty

0    True
1    True
2    True
3    True
4    True
5    True
6    True
7    True
8    True
9    True
Name: score, dtype: bool


,id,name,subjects,score,subject
0,1,Alice,Math,85,Math
1,2,Bob,Science,92,Science
2,3,Charlie,English,78,English
3,4,David,Computer,88,Computer
4,5,Emma,Physics,95,Physics
5,6,Frank,Chemistry,81,Chemistry
6,7,Grace,Biology,76,Biology
7,8,Henry,History,89,History
8,9,Isabella,Economics,91,Economics
9,10,Jack,Geography,84,Geography


In [81]:
df.isna().sum() #Counts them per column
df.dropna() #drops rows with any NaN
df.dropna(subset=["Suppressed"])
df["Data_value"]=df["Data_value"].fillna(0)
df.groupby("Data_value").get_group(0).head()



,Series_reference,Period,Data_value,Suppressed,STATUS,UNITS,Magnitude,Subject,Group,Series_title_1,Series_title_2,Series_title_3,Series_title_4,Series_title_5
92,BDCQ.SEA1AT,2011.06,0.0,NaN,C,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Trend,NaN,NaN
93,BDCQ.SEA1AT,2011.09,0.0,NaN,C,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Trend,NaN,NaN
94,BDCQ.SEA1AT,2011.12,0.0,NaN,C,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Trend,NaN,NaN
95,BDCQ.SEA1AT,2012.03,0.0,NaN,C,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Trend,NaN,NaN
96,BDCQ.SEA1AT,2012.06,0.0,NaN,C,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Trend,NaN,NaN


In [91]:

#Fill empty with mean
df["Data_value"]=df["Data_value"].fillna(df["Data_value"].mean())
df.groupby("Data_value").get_group(df["Data_value"].mean()).head()

,Series_reference,Period,Data_value,Suppressed,STATUS,UNITS,Magnitude,Subject,Group,Series_title_1,Series_title_2,Series_title_3,Series_title_4,Series_title_5
92,BDCQ.SEA1AT,2011.06,83802.180055,NaN,C,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Trend,NaN,NaN
93,BDCQ.SEA1AT,2011.09,83802.180055,NaN,C,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Trend,NaN,NaN
94,BDCQ.SEA1AT,2011.12,83802.180055,NaN,C,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Trend,NaN,NaN
95,BDCQ.SEA1AT,2012.03,83802.180055,NaN,C,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Trend,NaN,NaN
96,BDCQ.SEA1AT,2012.06,83802.180055,NaN,C,Number,0,Business Data Collection - BDC,Industry by employment variable,Filled jobs,"Agriculture, Forestry and Fishing",Trend,NaN,NaN


In [102]:
print(df.columns.tolist())

['Series_reference', 'Period', 'Data_value', 'Suppressed', 'STATUS', 'UNITS', 'Magnitude', 'Subject', 'Group', 'Series_title_1', 'Series_title_2', 'Series_title_3', 'Series_title_4', 'Series_title_5']


In [134]:
print(df[df["Data_value"] > 80000]) #gives all the data
print(df[df["Data_value"] > 80000].shape[0]) #Gives total row for shape index 0


    Series_reference   Period  Data_value  Suppressed STATUS   UNITS  \
0        BDCQ.SEA1AA  2011.06     80078.0         NaN      F  Number   
2        BDCQ.SEA1AA  2011.12     85850.0         NaN      F  Number   
3        BDCQ.SEA1AA  2012.03     90743.0         NaN      F  Number   
4        BDCQ.SEA1AA  2012.06     81780.0         NaN      F  Number   
6        BDCQ.SEA1AA  2012.12     87793.0         NaN      F  Number   
..               ...      ...         ...         ...    ...     ...   
363      BDCQ.SEA1CS  2021.09    222510.0         NaN      R  Number   
364      BDCQ.SEA1CS  2021.12    222404.0         NaN      R  Number   
365      BDCQ.SEA1CS  2022.03    223712.0         NaN      R  Number   
366      BDCQ.SEA1CS  2022.06    224260.0         NaN      R  Number   
367      BDCQ.SEA1CS  2022.09    226410.0         NaN      F  Number   

     Magnitude                         Subject  \
0            0  Business Data Collection - BDC   
2            0  Business Data Colle

In [113]:
print(df.groupby("Series_reference")["Data_value"].mean())
#Groups different series refrence and find mean values of data for each group

Series_reference
BDCQ.SEA1AA     92620.704545
BDCQ.SEA1AS     92257.413043
BDCQ.SEA1CA    215282.869565
BDCQ.SEA1CS    215355.108696
Name: Data_value, dtype: float64


"""
               print(df.groupby("Data_value").get_group(83802.180055))

In Python, decimals often have tiny hidden precision differences in memory (e.g., 83802.180055000003).
get_group() requires an exact binary match, so tiny floating-point variations cause a KeyError.
sO USE NUMPY to get to the closest near value
"""

In [133]:
df= df.copy()
grouped = df.groupby("Data_value")
target_value = 83802.180055
exact_key = min(grouped.groups.keys(),key = lambda k : abs(float(k)-target_value))
result = grouped.get_group(exact_key)
print(result)

   Series_reference   Period  Data_value  Suppressed STATUS   UNITS  \
46      BDCQ.SEA1AS  2011.06     83971.0         NaN      R  Number   

    Magnitude                         Subject  \
46          0  Business Data Collection - BDC   

                              Group Series_title_1  \
46  Industry by employment variable    Filled jobs   

                       Series_title_2       Series_title_3  Series_title_4  \
46  Agriculture, Forestry and Fishing  Seasonally adjusted             NaN   

    Series_title_5  
46             NaN  


In [130]:
#Use of min unction and lambda
students = [
{"name": "Ana", "score": 88},
    {"name": "Ben", "score": 45},
    {"name": "Cid", "score": 72},

]
worst = min(students , key = lambda s : s["score"])
print(worst)


{'name': 'Ben', 'score': 45}


In [143]:
df.groupby("Subject").size()
#group by creates objects group soit doesnt have shape

Subject
Business Data Collection - BDC    499
dtype: int64

In [145]:
df["Data_value"].isna().sum()


np.int64(138)

In [149]:
len(pd.merge(students,scores,on="id"))

10